### PR TITLE
Apt pinned packages

### PR DESCRIPTION
**Pull request by dheyay**
## Why is this needed?

Fix for #3184
Documentation update PR: #3328 

## Test Steps

---

- [ ] *(un)check this to re-run the checklist action*

### CODE REVIEW

File: uaclient/security_status.py [hunk 1]
```diff
@@ -154,6 +157,7 @@ def filter_updates(
         for package in packages:
             # We only care about installed packages here
             if package.current_ver:
+                candidate_version = policy.get_candidate_ver(package)
```
Comment (renanrodrigo) [inline uaclient/security_status.py:160]: Sorry I know this is still a WIP.

I'd say this function is meant to filter the available updates for a package list, and the candidate logic should not live there.
I'd say the `create_updates_list` function is the one which should have knowledge about the candidates, so it can be passed to `get_update_status` and treated there as a separate case, instead of a separate variable drilled down.
I know depcaches have candidates for packages - I suggest a small helper function:
`_is_version_the_candidate(version: apt_pkg.Version) -> bool:`
which would check the parent package candidate and see it it matches the version, or something close to this.
Also, consider `lru_cache`-ing the logic which finds a candidate for a package, so you don't have to invoke the depcache multiple times for the same package (although this should be minimal / fine for this step)

File: uaclient/security_status.py [hunk 2]
```diff
@@ -175,24 +189,37 @@ def filter_updates(
...
                                 )
                                 if service:
+                                    upgrade_status = (
+                                        None
+                                        if version == esm_candidate_version
+                                        else UpdateStatus.AVAILABLE_LOWER_PRIORITY.value
+                                    )
                                     result[service].append(
-                                        (version, origin.site)
+                                        (version, origin.site, upgrade_status)
```
Comment (renanrodrigo) [inline uaclient/security_status.py:222]: When the user does not have a service enabled, the upgrade will never be available
This part is not needed afaics

File: uaclient/security_status.py (hunk 1 above)
Comment (dheyay) [reply uaclient/security_status.py:160]: Good cache (catch) :eyes: 
Yeah since we discussed not going the policy route, it makes sense to get a different function and not populate the filter function with this task.

File: uaclient/security_status.py [hunk 3]
```diff
@@ -118,15 +120,35 @@ def get_origin_for_installed_package(
...
 
 
-def get_update_status(service_name: str, ua_info: Dict[str, Any]) -> str:
+@lru_cache(maxsize=None)
+def _is_candidate_version(pkg: str, version: str) -> bool:
+    """Returns True if the package version is a candidate version."""
+    candidate_version = get_pkg_candidate_version(pkg, check_esm_cache=False)
+    if candidate_version:
+        return version == candidate_version
+    return False
```
Comment (lucasmoura) [inline uaclient/security_status.py:129]: Maybe we can replace this for a one liner:
`return version == candidate_version`

Or is there a situation where `version` is None too ?

**Review by lucasmoura — COMMENTED** [review]
@dheyay @renanrodrigo do you think we could come up with an integration test for this ?

File: uaclient/security_status.py (hunk 3 above)
Comment (dheyay) [reply uaclient/security_status.py:129]: The `get_pkg_candidate_version` function has the return type of `Optional[str]` so I believe there is a case where the candidate version returned can be `None`

**Review by orndorffgrant — CHANGES_REQUESTED** [review]
As part of our process to get ready for v35, I have rebased `next-v35` onto `main`. As a result, `next-v35` is now obsolete. Please rebase this on `main` and target the PR to `main`.

File: uaclient/security_status.py [hunk 4]
```diff
@@ -256,20 +258,40 @@ def get_livepatch_fixed_cves() -> List[Dict[str, Any]]:
     return []
 
 
+@lru_cache(maxsize=None)
+def _is_candidate_version(pkg: str, version: str) -> bool:
+    """Returns True if the version is a candidate version."""
+    candidate_version = get_pkg_candidate_version(pkg, check_esm_cache=True)
+    # print(f"candidate_version: {candidate_version}, version: {version}")
```
[the remark this answers is not in the document]
Comment (orndorffgrant) [reply uaclient/security_status.py]: I didn't even notice this was a stale review comment there - sorry.

[the remark this answers is not in the document]
Comment (dheyay) [reply uaclient/security_status.py]: Removed this while editing currently, should be good, thanks :smile:

File: uaclient/security_status.py [hunk 5]
```diff
@@ -118,15 +120,35 @@ def get_origin_for_installed_package(
...
+    if service_name in ("standard-security", "standard-updates"):
+        is_candidate = _is_candidate_version(
+            version.parent_pkg.name, version.ver_str
+        )
+        return (
+            UpdateStatus.AVAILABLE.value
+            if is_candidate
+            else UpdateStatus.AVAILABLE_NOT_PREFERRED.value
+        )
+    if ua_info["attached"] and service_name in ua_info["enabled_services"]:
```
Comment (renanrodrigo) [inline uaclient/security_status.py:151]: Is there a particular reason to apply the logic only to the archive, but not to possible underpreferred ESM versions?
I mean, even if those are not a planned thing, they may happen and the criteria is the same - anything available but not a candidate should be in the new state

File: uaclient/security_status.py (hunk 3 above)
Comment (renanrodrigo) [reply uaclient/security_status.py:129]: Candidate will be None when you request a package that is not in the Cache, or a package which is in the Cache but has no installable version (or is pinned to never, or things like that) so yeah for the sake of corner cases, keep the check

**Comment by renanrodrigo** [comment]
Writing the integration test would be hard (as you probably figured out by now) because the situation where we have unpreferred updates is (or should be) temporary (see the issue - the package which triggered it is fixed already). Either we fake it all (no reason to have an integration test then, just the unit test will do) or we reason about it and move on.

In the second case, I would still go for a manual test, even if i involves manipulating apt caches, just to make sure.

File: uaclient/security_status.py (hunk 5 above)
Comment (dheyay) [reply uaclient/security_status.py]: Updated this as per standup discussion :smile:

File: uaclient/security_status.py [hunk 6]
```diff
@@ -118,7 +120,21 @@ def get_origin_for_installed_package(
     return "third-party"
 
 
-def get_update_status(service_name: str, ua_info: Dict[str, Any]) -> str:
+@lru_cache(maxsize=None)
+def _is_candidate_version(pkg: str, version: str) -> bool:
+    """Returns True if the package version is a candidate version."""
+    candidate_version = get_pkg_candidate_version(pkg, check_esm_cache=False)
+    print(pkg, " ", version, " ", candidate_version)
```
Comment (orndorffgrant) [inline uaclient/security_status.py:127]:
```suggestion
```

**Review by orndorffgrant — COMMENTED** [review]
Thanks this looks good! Approval except for below:

From the checklist:

> - [x] No documentation updates are necessary for this change

This introduces a new enum value, which must at least be described here: https://canonical-ubuntu-pro-client.readthedocs-hosted.com/en/latest/explanations/how_to_interpret_the_security_status_command/#u-pro-packages-updates-v1

**Review by lucasmoura — APPROVED** [review]

**Review by orndorffgrant — APPROVED** [review]
